docs/rfcs/0005-ecommerce-rest-api.md

```diff
@@ -0,0 +1,182 @@
+
+One principle guiding these REST APIs is that our web server should never handle credit card numbers
+or other similarly sensitive information. Instead this information will be communicated
+directly to CyberSource from the frontend, or left out entirely so that the user enters
+this information on CyberSource's form.
+
+### /api/products/
+
+This API provides a view into what products are available for a user to purchase. This API
+will be used by product listing pages but won't be needed on the checkout page. The same
+information will be available via the basket REST API.
+
+#### GET
+
+A response would look like
+
+    [
+        {
+            "id": 3,  // this is ProductVersion.id, not Product.id
+            "description": "Supply Chain Management",
+            "price": "67.89"
+        },
+        ...
+    ]
+
+
+### /api/basket/
+
+This API allows users to interact with their basket to get its status, to update the items
+to purchase in the basket, and to add on a coupon.
+
+#### GET
+
+A response would look like
+
+    {
+        "items": [
+            {
+                "id": 3,  // this is ProductVersion.id, not Product.id
+                "description": "Supply Chain Management",
+                "price": "67.89"
+            }
+        ],
+        "coupons": [
+            {
+                "code": "DISCOUNT_15_OFF",
+                "amount": "0.15",
+                "targets": [3]  // ProductVersion.id. There may be more but products not already in cart will be filtered out.
+            }
+        ]
+    }
+
+An empty basket would return:
+
+    {
+        "items": [],
+        "coupons": []
+    }
+
+#### PATCH
+
+A user can PATCH the basket to update the items list or to attempt to use a coupon. Each
+time a user PATCHes this endpoint, the basket will be validated. If the basket is not valid
+a 400 error will be returned and the basket will be left unchanged.
+
+Also note that a basket passing validation is not a guarantee that the checkout will
+pass validation. This could happen if the coupon code was valid when it was being
+added to the basket but it was not valid by the time of checkout. A product may also
+have its validity changed, preventing checkout from working.
+
+TODO: what happens if a basket becomes invalid over time? I think this is somewhat rare
+but we should figure out how to handle it. We can just clear the basket but I don't know when
+we should do this validation.
+
+##### Clear a basket
+
+A user could clear a basket by patching:
+
+    {
+        "items": [],
+        "coupons": []
+    }
+    
+They could clear just the items in the basket:
+
+    {
+        "items": []
+    }
+    
+However this would leave the coupon code as is. If there was a coupon code it would
+fail validation because there is nothing to apply the coupon to, and the basket would
+be left unchanged. So there is no use case for this.
+
+A user could clear only the coupon:
+
+    {
+        "coupons": []
+    }
```

Comment:
If a user manually applies a coupon code, then clears that coupon, should the best available automatic coupon for the basket product be applied again?

Reply:
Yes I think that's right. I'll add some text to make that explicit.